cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -66,22 +67,29 @@ func (e eventTypeSet) matches(event xdr.ContractEvent) bool {
 }
 
 type EventInfo struct {
-	EventType                string   `json:"type"`
-	Ledger                   int32    `json:"ledger"`
-	LedgerClosedAt           string   `json:"ledgerClosedAt"`
-	ContractID               string   `json:"contractId"`
-	ID                       string   `json:"id"`
-	PagingToken              string   `json:"pagingToken"`
-	Topic                    []string `json:"topic"`
-	Value                    string   `json:"value"`
-	InSuccessfulContractCall bool     `json:"inSuccessfulContractCall"`
-	TransactionHash          string   `json:"txHash"`
+	EventType                string `json:"type"`
+	Ledger                   int32  `json:"ledger"`
+	LedgerClosedAt           string `json:"ledgerClosedAt"`
+	ContractID               string `json:"contractId"`
+	ID                       string `json:"id"`
+	PagingToken              string `json:"pagingToken"`
+	InSuccessfulContractCall bool   `json:"inSuccessfulContractCall"`
+	TransactionHash          string `json:"txHash"`
+
+	// TopicXDR is a base64-encoded list of ScVals
+	TopicXDR  []string                 `json:"topic,omitempty"`
+	TopicJSON []map[string]interface{} `json:"topicJson,omitempty"`
+
+	// ValueXDR is a base64-encoded ScVal
+	ValueXDR  string                 `json:"value,omitempty"`
+	ValueJSON map[string]interface{} `json:"valueJson,omitempty"`
```

Comment:
💡 Given that the JSON is encoded by the Rust, I assume the presence of a map here means that the Go is first decoding the Rust JSON then re-encoding that here.

I suggest you replace this map with json.RawMessage ([godoc](https://pkg.go.dev/encoding/json#RawMessage)) which allows you to inject the precomputed JSON from the Rust directly in without having to do the encode-decode-encode dance.

Reply:
Yes, you're right that the Rust code was returning a JSON string and that the Go code was giving that structure via `json.Unmarshal`. While it's a brilliant idea to just inject it directly, I'm worried the JSON serialization of the response will just treat it like a string? I'll have to try it out. Also, it unfortunately prevents any inspection of the structure (I am checking for an `error` field as a way to transmit errors across the FFI).

Reply:
Done, brilliant :pray: